### PR TITLE
Support blob transaction type (EIP-4844)

### PR DESCRIPTION
**Pull request by rodiazet**
- Add blob tx type support in state. 
- Modify `transition` and `validate_transaction` functions to pass `blob_gas_left` parameter.
- Update unit tests
- Support blob tx in t8n
- Compute `excess_blob_gas` and `blob_gas_price`.

### CODE REVIEW

File: test/state/state.hpp [hunk 1]
```diff
@@ -184,11 +200,11 @@ void finalize(State& state, evmc_revision rev, const address& coinbase,
 
 [[nodiscard]] std::variant<TransactionReceipt, std::error_code> transition(State& state,
     const BlockInfo& block, const Transaction& tx, evmc_revision rev, evmc::VM& vm,
-    int64_t block_gas_left);
+    int64_t block_gas_left, uint64_t blob_gas_left);
```
Comment (gumb0) [inline test/state/state.hpp:203]: Why did you make `blob_gas_left` unsigned, when `block_gas_left` is signed?

Comment (rodiazet) [reply test/state/state.hpp]: fixed. No idea actually.

**Review by chfast — COMMENTED** [review]
How was this tested? I think we need to enable at least some external JSON tests on CI.

**Comment by rodiazet** [comment]
> How was this tested? I think we need to enable at least some external JSON tests on CI.

Done by https://github.com/ethereum/evmone/pull/713/commits/3dfcf0de9ea4acbe686c826f5575698e096ab08c

File: test/state/state.hpp [hunk 2]
```diff
@@ -165,6 +178,9 @@ struct TransactionReceipt
     /// Amount of gas used by this transaction.
     int64_t gas_used = 0;
 
+    /// Amount of blob gas used by this transaction.
+    int64_t blob_gas_used = 0;
```
Comment (chfast) [inline test/state/state.hpp:182]: This is always constant times number of blob hashes. Maybe replace this with a member `blob_gas_used() { GAS_PER_BLOB * blob_hashes.size() }`?

File: test/state/state.hpp [hunk 3]
```diff
@@ -99,6 +99,13 @@ struct BlockInfo
     bytes32 prev_randao;
     uint64_t base_fee = 0;
     std::vector<Ommer> ommers = {};
+
+    /// The "excess blob gas" parameter from EIP-4844
+    /// for computing the blob gas price in the current block.
+    int64_t excess_blob_gas = 0;
+    /// Max amount of blob gas allowed in block. It's static now but can be dynamic in the future.
+    static constexpr int64_t max_blob_gas_per_block = 786432;
```
Comment (chfast) [inline test/state/state.hpp:107]:
```suggestion
    static constexpr int64_t MaxBlobGasPerBlock = 786432;
```

File: test/state/state.cpp [hunk 4]
```diff
@@ -307,6 +368,20 @@ std::variant<TransactionReceipt, std::error_code> transition(State& state, const
                    tx.to.has_value() ? tx.to.value() : bytes_view(), tx.value, tx.data,
                    tx.access_list, static_cast<bool>(tx.v), tx.r, tx.s);
     }
+    else
+    {
+        if (tx.v > 1)
+            throw std::invalid_argument("`v` value for eip1559 transaction must be 0 or 1");
```
Comment (chfast) [inline test/state/state.cpp:374]:
```suggestion
            throw std::invalid_argument("`v` value for blob transaction must be 0 or 1");
```

File: test/state/state.hpp (hunk 2 above)
Comment (rodiazet) [reply test/state/state.hpp:182]: We don't have access to `blob_hashes.size()` in the context of `TransactionReceipt`  struct.

File: test/state/state.cpp [hunk 5]
```diff
@@ -49,8 +65,10 @@ int64_t compute_tx_intrinsic_cost(evmc_revision rev, const Transaction& tx) noex
     const auto initcode_cost =
         is_create && rev >= EVMC_SHANGHAI ? initcode_word_cost * num_words(tx.data.size()) : 0;
     const auto tx_cost = is_create && rev >= EVMC_HOMESTEAD ? create_tx_cost : call_tx_cost;
-    return tx_cost + compute_tx_data_cost(rev, tx.data) + compute_access_list_cost(tx.access_list) +
-           initcode_cost;
+    auto c = tx_cost + compute_tx_data_cost(rev, tx.data) +
```
[the remark this answers is not in the document]
Comment (rodiazet) [reply test/state/state.cpp]: fixed

File: test/state/state.cpp [hunk 6]
```diff
@@ -75,11 +93,32 @@ evmc_message build_message(const Transaction& tx, int64_t execution_gas_limit) n
...
+            return make_error_code(EMPTY_BLOB_HASHES_LIST);
+        if (tx.blob_hashes.size() > 6)
+            return make_error_code(BLOB_HASHES_LIST_SIZE_LIMIT_EXCEEDED);
+
+        if (tx.max_blob_gas_price < blob_gas_price)
+            return make_error_code(FEE_CAP_LESS_THEN_BLOCKS);
+
+        if (std::ranges::any_of(tx.blob_hashes, [](const auto& h) { return h.bytes[0] != 0x01; }))
+            return make_error_code(INVALID_BLOB_HASH_VERSION);
+        if (GAS_PER_BLOB * tx.blob_hashes.size() > static_cast<uint64_t>(blob_gas_left))
```
[the remark this answers is not in the document]
Comment (rodiazet) [reply test/state/state.cpp]: done

File: test/state/state.cpp [hunk 7]
```diff
@@ -307,6 +368,20 @@ std::variant<TransactionReceipt, std::error_code> transition(State& state, const
...
+    {
+        if (tx.v > 1)
+            throw std::invalid_argument("`v` value for eip1559 transaction must be 0 or 1");
+        // tx_type +
+        // rlp [chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit, to, value,
+        // data, access_list, max_fee_per_blob_gas, blob_versioned_hashes, sig_parity, r, s];
+        return bytes{stdx::to_underlying(Transaction::Type::blob)} +
+               rlp::encode_tuple(tx.chain_id, tx.nonce, tx.max_priority_gas_price, tx.max_gas_price,
+                   static_cast<uint64_t>(tx.gas_limit),
+                   tx.to.has_value() ? tx.to.value() : bytes_view(), tx.value, tx.data,
```
[the remark this answers is not in the document]
Comment (rodiazet) [reply test/state/state.cpp]: done

File: test/state/state.hpp (hunk 2 above)
Comment (gumb0) [reply test/state/state.hpp:182]: You should probably get this value form a transaction, not from receipt. This is not part of receipt as defined in EIP.

File: test/state/state.hpp [hunk 8]
```diff
@@ -123,17 +130,23 @@ struct Transaction
...
+        /// Introduced by EIP-4844 https://eips.ethereum.org/EIPS/eip-4844.
+        blob = 3,
     };
 
     Type type = Type::legacy;
     bytes data;
     int64_t gas_limit;
     intx::uint256 max_gas_price;
     intx::uint256 max_priority_gas_price;
+    intx::uint256 max_blob_gas_price = {};
```
Comment (gumb0) [inline test/state/state.hpp:144]: redundant
```suggestion
    intx::uint256 max_blob_gas_price;
```

File: test/state/state.hpp [hunk 9]
```diff
@@ -123,17 +130,23 @@ struct Transaction
...
     address sender;
     std::optional<address> to;
     intx::uint256 value;
     AccessList access_list;
+    std::vector<bytes32> blob_hashes = {};
```
Comment (gumb0) [inline test/state/state.hpp:149]:
```suggestion
    std::vector<bytes32> blob_hashes;
```

File: test/state/state.cpp [hunk 10]
```diff
@@ -75,11 +91,32 @@ evmc_message build_message(const Transaction& tx, int64_t execution_gas_limit) n
...
     switch (tx.type)
     {
+    case Transaction::Type::blob:
+        if (rev < EVMC_CANCUN)
+            return make_error_code(TX_TYPE_NOT_SUPPORTED);
+        if (!tx.to.has_value())
+            return make_error_code(CREATE_BLOB_TX);
+        if (tx.blob_hashes.empty())
+            return make_error_code(EMPTY_BLOB_HASHES_LIST);
+        if (tx.blob_hashes.size() > 6)
```
Comment (gumb0) [inline test/state/state.cpp:108]: I understand that this is a limit for number of blob hashes included in a block, and transaction validity doesn't include this?

File: test/state/state.cpp [hunk 11]
```diff
@@ -75,11 +91,32 @@ evmc_message build_message(const Transaction& tx, int64_t execution_gas_limit) n
 /// Validates transaction and computes its execution gas limit (the amount of gas provided to EVM).
 /// @return  Execution gas limit or transaction validation error.
 std::variant<int64_t, std::error_code> validate_transaction(const Account& sender_acc,
-    const BlockInfo& block, const Transaction& tx, evmc_revision rev,
-    int64_t block_gas_left) noexcept
+    const BlockInfo& block, const Transaction& tx, evmc_revision rev, int64_t block_gas_left,
+    int64_t blob_gas_left) noexcept
 {
+    const auto blob_gas_price = fake_exponential(1, block.excess_blob_gas, 3338477);
```
Comment (pdobacz) [inline test/state/state.cpp:97]: let's define these 2 constants somewhere

**Review by pdobacz — COMMENTED** [review]
I need some more time to mull this over, but two comments from me so far

File: test/state/state.hpp (hunk 2 above)
Comment (rodiazet) [reply test/state/state.hpp:182]: OK. I get Pawel's comment now. Thanks.

File: test/state/state.cpp (hunk 10 above)
Comment (rodiazet) [reply test/state/state.cpp:106]: Yeah. IMO it's already handled by the condition but maybe @chfast wants to take a look also. I don't remember why this additional `BLOB_HASHES_LIST_SIZE_LIMIT_EXCEEDED` error was added. Maybe it's not needed any longer. I would remove it now.

Comment (chfast) [reply test/state/state.cpp:106]: If these checks are equivalent we can remove the `tx.blob_hashes.size() > 6`. I only recall discussions about having 3/6 or 2/4 blobs in the block. I'm not sure if blob gas cost handles this exact limit precisely and is enough.

Comment (chfast) [reply test/state/state.cpp:106]: Anyway, if the EIP does not mention 6 at all we proceed without it and check against tests.

Comment (rodiazet) [reply test/state/state.cpp:106]: `6*0x20000 == 786432` which is precisely `MAX_BLOB_GAS_PER_BLOCK`

File: test/state/state.cpp (hunk 11 above)
Comment (chfast) [reply test/state/state.cpp:95]: Do you think we should go with Python naming convention for constants to be consistent with the spec? (sent from CLion PR review)

Comment (rodiazet) [reply test/state/state.cpp:95]: BTW. We should have some common place to define constants like those two.

**Review by pdobacz — APPROVED** [review]

File: test/state/state.cpp [hunk 12]
```diff
@@ -14,6 +14,10 @@ namespace evmone::state
 {
 namespace
 {
+/// Values specified by EIP-4844
+constexpr auto min_blob_gasprice = 1;
```
Comment (chfast) [inline test/state/state.cpp:18]: Bring the constants to the single place they are use, which is `if (tx.type == Transaction::Type::blob)`.
Use naming `MIN_BLOB_GASPRICE` or `MinBlobGasPrice`.
They will be still searchable but no need to extend their scope unnecessarily.

File: test/state/state.cpp [hunk 13]
```diff
@@ -307,6 +365,21 @@ std::variant<TransactionReceipt, std::error_code> transition(State& state, const
                    tx.to.has_value() ? tx.to.value() : bytes_view(), tx.value, tx.data,
                    tx.access_list, static_cast<bool>(tx.v), tx.r, tx.s);
     }
+    else
+    {
+        if (tx.v > 1)
+            throw std::invalid_argument("`v` value for blob transaction must be 0 or 1");
+        if (!tx.to.has_value())  // Blob tx has to have `to` address
+            throw std::invalid_argument("`to` value for blob transaction must not be null");
```
Comment (chfast) [inline test/state/state.cpp:373]: Codecov says this is uncovered by tests. Please add a test if true.

**Review by chfast — CHANGES_REQUESTED** [review]

File: test/state/state.cpp (hunk 13 above)
Comment (rodiazet) [reply test/state/state.cpp:373]: Test addded

**Comment by chfast** [comment]
These tests from execution-spec-tests are failing: `cancun/eip4844_blobs/excess_blob_gas`.

File: test/statetest/statetest_loader.cpp [hunk 14]
```diff
@@ -147,6 +147,29 @@ inline uint64_t calculate_current_base_fee_eip1559(
     return base_fee;
 }
 
+intx::uint256 compute_blob_gas_price(uint64_t excess_blob_gas) noexcept
```
Comment (chfast) [inline test/statetest/statetest_loader.cpp:150]: This should stay in `state.cpp` because we want to have unit tests for it.

**Review by chfast — CHANGES_REQUESTED** [review]

**Review by chfast — APPROVED** [review]

**Pull request merged**